### PR TITLE
Issue 338: SelfTest using Pravega Client & external clusters.

### LINKED ISSUE #338: Pravega SelfTest:  Multi-instances
Enhance SelfTest to have the following abilities:
- Target a Pravega Cluster, in addition to a single SegmentStore instance.
    - Fully implement Pravega Client integration - this involves reads and transactions as well.
    - Extend the current Tail/Catchup reads to work with Streams (and not just Segments) and verify ordering of data within a particular Key. Note that Tier2 reads will be close to impossible to perform due to the fact that data with different keys are mingled together in the same segment.

### PR DESCRIPTION
**Pull request by andreipaduroiu**
**Change log description**
Reworked SelfTest to be able to target an entire Pravega Cluster, in addition to just one SegmentStore instance.

**Overview**:
- **Accepted inputs**: a large variety, including # Streams, # Producers, Producer Parallelism, # Controllers, # Segment Stores, etc.
- **Supported operations**: Append, Seal, Transactions, Read, Create & Delete.
- `Producers` generate `Event`s which are made up of a header, containing the originator id, a routing key, a sequence number and some random payload.
- `Consumers` (one per Stream), do tail reads and random-address reads, and validate that the Events are read back in order and are not corrupted - this is **quality verification**.
- **Performance check**: reports various stats on all executed operations, such as latency and throughput (instant/overall).

**Types of tests**:
- **SegmentStoreDirect**: instantiates a single SegmentStore in-process, against either an in-memory Tier1 or BK as Tier1, and an in-memory, truncateable Tier2. Verifies all features supported by the SegmentStore.
- **InProcessMockListener**: instantiates a dud SegmentStore (all operations are instant) and uses a `PravegaConnectionListener` and `ClientFactory` to execute append operations to it.  No other operations are supported.
    - This is used mainly for client performance checks
- **InProcessStoreListener**: instantiates a single SegmentStore, a `PravegaConnectionListener` and uses a real client to execute append operations to it. No other operations are supported.
- **External**: Connects to an external (existing) Pravega cluster made up of a Controller and SegmentStore and uses the Client to execute appends, reads, seals and transactions.
- **OutOfProcess**: same as External, but it auto-creates the Pravega cluster on the local machine, with a specified number of SegmentStores, Bookies, Controllers, each in their own process.
- **BookKeeper**: executes appends against an out-of-process BookKeeper instance (with a custom number of bookies and replication factor). Each stream is one Ledger.

Also added a gradle task for this so it can be run from the command line.

**Purpose of the change**
Fixes #338. A comprehensive end-to-end test that can be used to target various parts of the Pravega Cluster.

**What the code does**
No changes to main production code or unit tests (except for 1-2 methods changed to enhance their usage).

**How to verify it**
`./gradlew test:integration:selftest`
Also created a Gradle app which makes a `tar` file for this that can be deployed.

### CODE REVIEW

File: common/src/main/java/io/pravega/common/lang/ProcessStarter.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import lombok.val;
+
+/**
+ * Helps start a Class out of Process.
+ */
+public class ProcessStarter {
```
Comment (tkaitchuck) [inline common/src/main/java/io/pravega/common/lang/ProcessStarter.java]: This doesn't look thread safe.

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperServiceRunner.java [hunk 2]
```diff
@@ -169,4 +190,32 @@ private void cleanupDirectories() throws IOException {
     }
 
     //endregion
+
+    public static void main(String[] args) throws Exception {
```
Comment (tkaitchuck) [inline segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperServiceRunner.java]: It's a little weird to add a main to a class that didn't used to have one and looks to be used already. Why is this here and not in a new file?

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperServiceRunner.java [hunk 3]
```diff
@@ -169,4 +190,32 @@ private void cleanupDirectories() throws IOException {
...
+            System.out.println(String.format("Invalid or missing arguments (via system properties). Expected: %s(int), " +
+                            "%s(int), %s(int), %s(String). (%s).", PROPERTY_BASE_PORT, PROPERTY_BOOKIE_COUNT, PROPERTY_ZK_PORT,
+                    PROPERTY_LEDGERS_PATH, ex.getMessage()));
+            System.exit(-1);
+            return;
+        }
+
+        BookKeeperServiceRunner runner = b.build();
+        runner.startAll();
+        Thread.sleep(Long.MAX_VALUE);
```
Comment (tkaitchuck) [inline segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperServiceRunner.java]: Should this be waiting on something like a future?

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ZooKeeperServiceRunner.java [hunk 4]
```diff
@@ -63,7 +64,34 @@ public void start() throws Exception {
...
+     *
+     * @param zkPort The ZooKeeper Port.
+     * @return True if ZooKeeper started within a specified timeout, false otherwise.
+     */
+    public static boolean waitForServerUp(int zkPort) {
+        val address = LOOPBACK_ADDRESS.getHostAddress() + ":" + zkPort;
+        return LocalBookKeeper.waitForServerUp(address, LocalBookKeeper.CONNECTION_TIMEOUT);
+    }
+
+    public static void main(String[] args) throws Exception {
```
Comment (tkaitchuck) [inline segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ZooKeeperServiceRunner.java]: Same comment as above.

File: test/integration/src/main/java/io/pravega/test/integration/selftest/Consumer.java [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.SneakyThrows;
+import lombok.val;
+
+/**
+ * Represents an operation consumer. Attaches to a StoreAdapter, listens to tail-reads (on segments), and validates
+ * incoming data. Currently this does not do catch-up reads or storage reads/validations.
+ */
+public class Consumer extends Actor {
```
Comment (tkaitchuck) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/Consumer.java]: Is there a way to convince git this is a move so that we can review a diff?

File: test/integration/src/main/java/io/pravega/test/integration/selftest/Event.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    //region Members
+
+    private static final int PREFIX_LENGTH = Integer.BYTES;
+    private static final int OWNER_ID_LENGTH = Integer.BYTES;
+    private static final int ROUTING_KEY_LENGTH = Integer.BYTES;
+    private static final int SEQUENCE_LENGTH = Integer.BYTES;
+    private static final int START_TIME_LENGTH = Long.BYTES;
+    private static final int LENGTH_LENGTH = Integer.BYTES;
+    static final int HEADER_LENGTH = PREFIX_LENGTH + OWNER_ID_LENGTH + ROUTING_KEY_LENGTH + SEQUENCE_LENGTH + START_TIME_LENGTH + LENGTH_LENGTH;
+    private static final int PREFIX = (int) Math.pow(Math.E, 20);
```
Comment (tkaitchuck) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/Event.java]: What!?

File: test/integration/src/main/java/io/pravega/test/integration/selftest/Event.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+import io.pravega.common.util.ByteArraySegment;
+import io.pravega.test.integration.selftest.adapters.TruncateableArray;
+import java.io.IOException;
+import lombok.Getter;
+import lombok.SneakyThrows;
+
+/**
+ * Represents an Event with a Routing Key and payload.
+ */
+public class Event {
```
Comment (tkaitchuck) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/Event.java]: A diff would also help here.

File: test/integration/src/main/java/io/pravega/test/integration/selftest/Consumer.java (hunk 5 above)
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/Consumer.java]: There are so many changes in this file I think it's best if you don't try a diff. I mostly rewrote the whole thing.

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperServiceRunner.java (hunk 3 above)
Comment (andreipaduroiu) [inline segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperServiceRunner.java]: There is no future to wait on here. This starts BK in the current process and exits when the process is killed. This is similar to how we start the Controller and SegmentStore services too.

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperServiceRunner.java (hunk 2 above)
Comment (andreipaduroiu) [inline segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperServiceRunner.java]: I added a comment explaining its purpose. It's so that I can start BK out of process.

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ZooKeeperServiceRunner.java (hunk 4 above)
Comment (andreipaduroiu) [inline segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ZooKeeperServiceRunner.java]: Added a comment explaining that the purpose of this is to start ZK out of process.

File: common/src/main/java/io/pravega/common/lang/ProcessStarter.java (hunk 1 above)
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/lang/ProcessStarter.java]: Added appropriate annotation.

File: build.gradle [hunk 8]
```diff
@@ -336,6 +342,13 @@ project('test:integration') {
         classpath = sourceSets.main.runtimeClasspath
         standardInput = System.in
     }
+
+    task selftest(type: JavaExec) {
+        main = "io.pravega.test.integration.selftest.SelfTestRunner"
+        classpath = sourceSets.main.runtimeClasspath
+        systemProperties System.getProperties()
+        jvmArgs "-Xms4g", "-Xmx16g"
```
Comment (andreipaduroiu) [inline build.gradle]: Oops. I forgot to remove that. I added it temporarily to debug something.

File: test/integration/src/main/java/io/pravega/test/integration/selftest/Event.java (hunk 7 above)
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/Event.java]: This is a new class. I moved some code over from another class, but I couldn't have renamed the other class since it was doing so many other things.

File: test/integration/src/main/java/io/pravega/test/integration/selftest/Event.java (hunk 6 above)
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/Event.java]: What? I wanted to have a specific arbitrary number that denotes the start of every event. If I get some byte stream back that doesn't start with that, then I know this can't be an event.

File: test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/BookKeeperAdapter.java [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+    private static final String LOG_ID = BookKeeperAdapter.class.getSimpleName();
+    private final TestConfig testConfig;
+    private final BookKeeperConfig bkConfig;
+    private final ScheduledExecutorService executor;
+    private final ConcurrentHashMap<String, DurableDataLog> logs;
+    @GuardedBy("internalIds")
+    private final HashMap<String, Integer> internalIds;
+    private final AtomicBoolean closed;
+    private final Thread stopBookKeeperProcess;
+    private Process bookKeeperService;
```
Comment (tkaitchuck) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/BookKeeperAdapter.java]: THese three either need to be guarded by a lock or this class should be made into a service, which will handle this for you (assuming they are initialized in doSstart)

File: test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/InProcessMockClientAdapter.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * Client-based adapter that targets an in-process Client with a Mock Controller and Mock StreamSegmentStore.
+ */
+class InProcessMockClientAdapter extends ClientAdapterBase {
+    //region Members
+    private static final String LOG_ID = "InProcessAdapter";
+    private static final String LISTENING_ADDRESS = "localhost";
+    private PravegaConnectionListener listener;
```
Comment (tkaitchuck) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/InProcessMockClientAdapter.java]: See my earlier comment

File: test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/SegmentStoreAdapter.java [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+    private final ScheduledExecutorService testExecutor;
+    private final TestConfig config;
+    private final ServiceBuilderConfig builderConfig;
+    private final AtomicBoolean closed;
+    private final AtomicBoolean initialized;
+    private final ServiceBuilder serviceBuilder;
+    private final AtomicReference<Storage> storage;
+    private final AtomicReference<ScheduledExecutorService> storeExecutor;
+    private final Thread stopBookKeeperProcess;
+    private Process bookKeeperService;
```
Comment (tkaitchuck) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/SegmentStoreAdapter.java]: Same here

File: test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/BookKeeperAdapter.java (hunk 9 above)
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/BookKeeperAdapter.java]: Fixed all the adapters to be Guava AbstractIdleService

File: test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/InProcessMockClientAdapter.java (hunk 10 above)
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/InProcessMockClientAdapter.java]: Fixed

File: test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/SegmentStoreAdapter.java (hunk 11 above)
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/SegmentStoreAdapter.java]: Fixed
